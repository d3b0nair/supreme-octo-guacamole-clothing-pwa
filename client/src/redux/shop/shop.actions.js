import ShopActionsTypes from "./shop.types";

import { collection, getDocs } from "firebase/firestore";

import {
  db,
  convertCollectionsSnapshotToMap,
} from "../../firebase/firebase.utils";

export const fetchCollectionsStart = () => ({
  type: ShopActionsTypes.FETCH_COLLECTIONS_START,
});

export const fetchCollectionsSuccess = (collectionsMap) => ({
  type: ShopActionsTypes.FETCH_COLLECTIONS_SUCCESS,
  payload: collectionsMap,
});

export const fetchCollectionsFailure = (errorMessage) => ({
  type: ShopActionsTypes.FETCH_COLLECTIONS_FAILURE,
  payload: errorMessage,
});
export const fetchCollectionsStartAsync = () => {
  return (dispatch) => {
    const collectionRef = collection(db, "collections");
    dispatch(fetchCollectionsStart());

    getDocs(collectionRef)
      .then((snapshot) => {
        const collectionsMap = convertCollectionsSnapshotToMap(snapshot);
        dispatch(fetchCollectionsSuccess(collectionsMap));
      })
      .catch((error) => dispatch(fetchCollectionsFailure(error.message)));
  };
};
