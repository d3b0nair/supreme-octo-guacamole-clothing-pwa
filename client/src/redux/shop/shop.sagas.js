import { takeLatest, call, put, all } from "redux-saga/effects";
import { collection, getDocs } from "firebase/firestore";

import {
  db,
  convertCollectionsSnapshotToMap,
} from "../../firebase/firebase.utils";

import {
  fetchCollectionsFailure,
  fetchCollectionsSuccess,
} from "./shop.actions";

import ShopActionsTypes from "./shop.types";

export function* fetchCollectionsAsync() {
  try {
    const collectionRef = collection(db, "collections");
    const snapshot = yield getDocs(collectionRef);
    const collectionsMap = yield call(
      convertCollectionsSnapshotToMap,
      snapshot
    );
    yield put(fetchCollectionsSuccess(collectionsMap));
  } catch (error) {
    yield put(fetchCollectionsFailure(error.message));
  }
}

export function* fetchCollectionsStart() {
  yield takeLatest(
    ShopActionsTypes.FETCH_COLLECTIONS_START,
    fetchCollectionsAsync
  );
}

export function* shopSagas() {
  yield all([call(fetchCollectionsStart)]);
}
