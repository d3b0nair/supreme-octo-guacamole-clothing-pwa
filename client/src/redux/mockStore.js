import { createStore } from "redux";
import { persistReducer } from "redux-persist";
import { rootReducer, persistConfig } from "./root-reducer";
export const createMockStore = ({ mockState, mockReducers }) => {
  const combinedReducers = persistReducer(
    persistConfig,
    rootReducer(mockReducers)
  );
  return createStore(combinedReducers, mockState);
};
