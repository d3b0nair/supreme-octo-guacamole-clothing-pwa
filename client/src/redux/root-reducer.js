import { combineReducers } from "redux";
import { persistReducer } from "redux-persist";
import storage from "redux-persist/lib/storage";

import userReducer from "./user/user.reducer";
import cartReducer from "./cart/cart.reducer";
import directoryReducer from "./directory/directory.reducer";
import shopReducer from "./shop/shop.reducer";
import checkoutReducer from "./checkout/checkout.reducer";
import stripeReducer from "./stripe/stripe.reducer";
import orderReducer from "./order/order.reducer"

const persistConfig = {
  key: "root",
  storage,
  whitelist: ["cart", "stripe"],
};
const rootReducer = combineReducers({
  user: userReducer,
  cart: cartReducer,
  checkout: checkoutReducer,
  directory: directoryReducer,
  shop: shopReducer,
  stripe: stripeReducer,
  selectedOrder: orderReducer
});
export default persistReducer(persistConfig, rootReducer);
