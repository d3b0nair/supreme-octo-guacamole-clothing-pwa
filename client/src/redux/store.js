import { createStore, applyMiddleware, compose } from "redux";
import { persistStore } from "redux-persist";
import logger from "redux-logger";
import createSagaMiddleware from "redux-saga";

import rootReducer from "./root-reducer";
import rootSaga from "./root-saga";

const sagaMiddleware = createSagaMiddleware();

const middlewares = [sagaMiddleware];
const composeEnhancers =
  (process.env.NODE_ENV !== "production" &&
    typeof window !== "undefined" &&
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) ||
  compose;
if (process.env.NODE_ENV === "development") {
  middlewares.push(logger);
}
export const store = createStore(
  rootReducer,
  composeEnhancers(applyMiddleware(...middlewares))
);

sagaMiddleware.run(rootSaga);

export const persistor = persistStore(store);
