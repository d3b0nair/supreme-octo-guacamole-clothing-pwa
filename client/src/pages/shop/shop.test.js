
import React from "react";
import { createMockStore } from "../../redux/mockStore";
import { ShopPage } from "./shop.component";
import { render, cleanup } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";

describe("ShopPage", () => {
  const mockReducer = (
    state = {
      isFetching: true,
    },
    action
  ) => state;

  const mockState = {
    shop: {
      isFetching: true,
    },
  };

  const mockFetchCollectionsStart = jest.fn();

  const store = createMockStore({
    state: mockState,
    reducers: { shop: mockReducer },
  });

  const mockMatch = {
    path: "",
  };

  const mockProps = (otherProps) => ({
    ...otherProps,
    match: mockMatch,
    fetchCollectionsStart: mockFetchCollectionsStart,
  });
  const component = (mockProps) => (
    <BrowserRouter>
      <Provider store={store}>
        <ShopPage {...mockProps} />
      </Provider>
    </BrowserRouter>
  );

  afterEach(cleanup);

  describe("Shop Page component", () => {
    it("should mount", () => {
      const { container } = render(component(mockProps()));
      expect(container).toMatchSnapshot();
    });
    it("should fetch collections", () => {
      render(component(mockProps()));
      expect(mockFetchCollectionsStart).toHaveBeenCalled();
    });
  });
});
