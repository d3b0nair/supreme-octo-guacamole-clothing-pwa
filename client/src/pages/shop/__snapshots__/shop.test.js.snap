// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShopPage Shop Page component should mount 1`] = `
<div>
  <div
    class="sc-dkPtRN hhzQVa"
  >
    <div
      class="sc-bdvvtL kHeCUE"
    >
      <div
        class="sc-gsDKAQ bMJKwb"
      />
    </div>
  </div>
</div>
`;
